### PR TITLE
HPCC-19607 ECLWatch index information page to include Blooms

### PR DESCRIPTION
**Pull request by kunalaswani**
Added new tab for Blooms information to Logical Files Details

Signed-off-by: Kunal Aswani <Kunal.Aswani@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [x] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by kunalaswani** [comment]
@GordonSmith please review

**Comment by kunalaswani** [comment]
HPCC-23109 was created for documentation.

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                            }
+                            return true;
+                        },
+                        sortable: false,
+                        unhidable: true
+                    }),
+                    Name: tree({
+                        width: 200,
+                        collapseOnRefresh: false,
+                        label: "Field Names",
```
Comment (miguelvazq) [inline esp/src/eclwatch/FileBloomsWidget.js]: Translation needed.

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                        sortable: false,
+                        unhidable: true
+                    }),
+                    FieldMaskName: { label: this.i18n.FieldMask, width: 100, sortable: false },
+                    LimitName: { label: this.i18n.Limit, width: 100, sortable: false },
+                    ProbabilityName: { label: this.i18n.Probability, width: 100, sortable: false }
+                }
+            }, domID);
+
+            retVal.on(".dgrid-row-url:click", function (evt) {
```
Comment (miguelvazq) [inline esp/src/eclwatch/FileBloomsWidget.js]: Are you actually opening anything?

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            });
+
+            return retVal;
+        },
+
+        _onRefresh: function () {
+            this.refreshGrid();
+        },
+
+        _onOpen: function (event) {
```
Comment (miguelvazq) [inline esp/src/eclwatch/FileBloomsWidget.js]: Are you actually opening anything?

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                var newRows = [];
+                if (lang.exists("DFUInfoResponse.FileDetail.Blooms.DFUFileBloom", response)) {
+                    results = response.DFUInfoResponse.FileDetail.Blooms.DFUFileBloom;
+                }
+
+                arrayUtil.forEach(results, function (row, idx) {
+                    lang.mixin(row, {
+                        __hpcc_parentName: null,
+                        __hpcc_id: idx,
+                        Name: nlsHPCC.Blooms,
```
Comment (miguelvazq) [inline esp/src/eclwatch/FileBloomsWidget.js]: Do we have to loop through them? Seems that your parent is static.

**Review by miguelvazq — CHANGES_REQUESTED** [review]

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+                return;
+
+            this._refreshActionState();
+            this.refreshGrid();
+
+            ESPUtil.MonitorVisibility(this.gridTab, function (visibility) {
+                if (visibility) {
+                    context.refreshGrid();
+                }
+            });
```
Comment (GordonSmith) [inline esp/src/eclwatch/FileBloomsWidget.js]: I suspect there is no need to refresh on tab selection (does it ever change once the file has been written)?

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                return this.query({
+                    __hpcc_parentName: parent.__hpcc_id
+                }, options);
+            };
+            var retVal = new declare([ESPUtil.Grid(true, true)])({
+                store: this.store,
+                columns: {
+                    col1: selector({
+                        width: 10,
+                        selectorType: 'checkbox',
```
Comment (GordonSmith) [inline esp/src/eclwatch/FileBloomsWidget.js]: Minor, should be " (double quotes) for consistentcy

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                    if (row.FieldNames) {
+                        arrayUtil.forEach(row.FieldNames.Item, function (item, itemIndex) {
+                            newRows.push({
+                                __hpcc_parentName: idx,
+                                __hpcc_id: item + itemIndex,
+                                Name: item,
+                                FieldMaskName: row.FieldMask,
+                                ProbabilityName: row.Probability,
+                                LimitName: row.Limit,
+                                type: "item",
```
Comment (GordonSmith) [inline esp/src/eclwatch/FileBloomsWidget.js]: Trailing comma

File: esp/src/eclwatch/templates/LFDetailsWidget.html [hunk 8]
```diff
@@ -254,6 +254,8 @@ <h2>
             </div>
             <div id="${id}_FileHistory" title="${i18n.History}" data-dojo-props="delayWidget: 'FileHistoryWidget'" data-dojo-type="DelayLoadWidget">
             </div>
+            <div id="${id}_FileBlooms" title="${i18n.Blooms}" data-dojo-props="delayWidget: 'FileBloomsWidget'" data-dojo-type="DelayLoadWidget">
+                </div>
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/LFDetailsWidget.html]: Indentation

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by kunalaswani** [comment]
@GordonSmith I spoke with Richard on what he would like to see, I have made the suggested changes from you and miguel as well as what Richard mentioned.

File: esp/src/eclwatch/LFDetailsWidget.js [hunk 9]
```diff
@@ -96,6 +98,7 @@ define([
                 this.workunitWidget = registry.byId(this.id + "_Workunit");
                 this.dfuWorkunitWidget = registry.byId(this.id + "_DFUWorkunit");
                 this.fileHistoryWidget = registry.byId(this.id + "_FileHistory");
+                this.fileBloomsWidget = registry.byId(this.id + "_FileBlooms")
```
Comment (miguelvazq) [inline esp/src/eclwatch/LFDetailsWidget.js]: semicolon

**Review by miguelvazq — COMMENTED** [review]
Minor comment.

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                                return false;
+                            }
+                            return true;
+                        },
+                        sortable: false,
+                        unhidable: true
+                    }),
+                    Name:{
+                        width: 200,
+                        collapseOnRefresh: false,
```
Comment (miguelvazq) [inline esp/src/eclwatch/FileBloomsWidget.js]: Not needed if your not using tree.

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                        sortable: false
+                    },
+                    LimitName: { label: this.i18n.Limit, width: 100, sortable: false },
+                    ProbabilityName: { label: this.i18n.Probability, width: 100, sortable: false }
+                }
+            }, domID);
+
+            retVal.on(".dgrid-row:dblclick", function (evt) {
+                evt.preventDefault();
+                context.grid.refresh()
```
Comment (miguelvazq) [inline esp/src/eclwatch/FileBloomsWidget.js]: Minor: Missing semicolon

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+            }).then(function (response) {
+                var results = [];
+
+                if (lang.exists("DFUInfoResponse.FileDetail.Blooms.DFUFileBloom", response)) {
+                    arrayUtil.forEach(response.DFUInfoResponse.FileDetail.Blooms.DFUFileBloom, function (row, idx) {
+                        lang.mixin(row, {
+                            Name: row.FieldNames.Item,
+                            ProbabilityName: row.Probability,
+                            LimitName: row.Limit,
+                            type: "fieldname"
```
Comment (miguelvazq) [inline esp/src/eclwatch/FileBloomsWidget.js]: Do you still need this type? Usually add these for renderCell methods to perform an action on those rows.

File: esp/src/eclwatch/nls/hpcc.js [hunk 13]
```diff
@@ -260,6 +261,8 @@ define({root:
     ExpandAll: "Expand All",
     Export: "Export",
     ExportSelectionsToList: "Export Selections to List",
+    FieldMask: "Field Mask",
```
Comment (miguelvazq) [inline esp/src/eclwatch/nls/hpcc.js]: FieldMask column is gone now.

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+        createGrid: function (domID) {
+            var context = this;
+            var retVal = new declare([ESPUtil.Grid(true, true)])({
+                store: this.store,
+                columns: {
+                    col1: selector({
+                        width: 10,
+                        selectorType: "checkbox",
+                        disabled: function (item) {
```
Comment (miguelvazq) [inline esp/src/eclwatch/FileBloomsWidget.js]: Probably do not need to disable anything since there is not actions to be taken.

**Review by miguelvazq — CHANGES_REQUESTED** [review]
One last minor change and then should be good.

**Review by miguelvazq — APPROVED** [review]

**Comment by kunalaswani** [comment]
@miguelvazq change made

**Comment by kunalaswani** [comment]
@GordonSmith changes were made, please review, thanks

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    "dojo/_base/array",
+
+    "dijit/registry",
+
+    "dgrid/selector",
+
+    "hpcc/GridDetailsWidget",
+    "src/WsDfu",
+    "src/ESPUtil",
+    "hpcc/DelayLoadWidget",
```
Comment (GordonSmith) [inline esp/src/eclwatch/FileBloomsWidget.js]: Trailing comma

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+                        firstLoad: true
+                    });
+                } else {
+                    currSel.init(currSel.params);
+                }
+            }
+        },
+
+        postCreate: function (args) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/FileBloomsWidget.js]: context not used

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by kunalaswani** [comment]
@GordonSmith changes made

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+    registry, selector,
+    GridDetailsWidget, WsDfu, ESPUtil, DelayLoadWidget) {
+    return declare("FileBloomsWidget", [GridDetailsWidget, ESPUtil.FormHelper], {
+        i18n: nlsHPCC,
+
+        gridTitle: nlsHPCC.title_Blooms,
+        idProperty: "__hpcc_id",
+
+        init: function (params) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/FileBloomsWidget.js]: `context` not used.

File: esp/src/eclwatch/FileBloomsWidget.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+            this._refreshActionState();
+            this.refreshGrid();
+        },
+
+        initTab: function () {
+            var currSel = this.getSelectedChild();
+            if (currSel && !currSel.initalized) {
+                if (currSel.id === this.id + "_Grid") {
+                    this.refreshGrid()
```
Comment (GordonSmith) [inline esp/src/eclwatch/FileBloomsWidget.js]: Missing semi-colon

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Two minor comments - otherwise looks good.

**Review by GordonSmith — APPROVED** [review]
Looks good...

**Pull request merged**